### PR TITLE
implemented order repository, order service methods

### PR DESCRIPTION
**Pull request by TemkaHillfinger**

### CODE REVIEW

File: src/main/java/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @PostMapping
+    public OrderDto placeAnOrder(@RequestBody String shippingAddress,
+                                 Authentication authentication) {
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        return orderService.saveOrder(userDetails.getUsername(), shippingAddress);
+    }
+
+    @GetMapping("/get-all")
```
Comment (MentorMA) [inline src/main/java/bookstore/controller/OrderController.java:32]: Check REST rules for endpoints

File: src/main/java/bookstore/dto/order/OrderDto.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import lombok.Data;
+
+@Data
+public class OrderDto {
+    private Long id;
+    private Long userId;
+    private Set<OrderItemDto> orderItems;
+    private LocalDateTime orderDate;
+    private BigDecimal total;
+    private Order.Status status;
```
Comment (MentorMA) [inline src/main/java/bookstore/dto/order/OrderDto.java:16]: What do you think about record?

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/controller/OrderController.java (hunk 1 above)
Comment (MAMentorFX) [reply src/main/java/bookstore/controller/OrderController.java:39]: image
```suggestion
    @GetMapping
```

File: src/main/java/bookstore/dto/order/OrderDto.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import java.time.LocalDateTime;
+import java.util.Set;
+
+public record OrderDto(Long id,
+         Long userId,
+         Set<OrderItemDto> orderItems,
+         LocalDateTime orderDate,
+         BigDecimal total,
+         Order.Status status){
+
```
Comment (MAMentorFX) [inline src/main/java/bookstore/dto/order/OrderDto.java:14]:
```suggestion
```

File: src/main/java/bookstore/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    public Order(ShoppingCart shoppingCart) {
+        this.user = shoppingCart.getUser();
+        this.status = Status.PENDING;
+        this.orderDate = LocalDateTime.now();
+    }
+
+    public Order() {
+
+    }
```
Comment (MAMentorFX) [inline src/main/java/bookstore/model/Order.java:57]: @NoArgsConstructor
```suggestion
```

File: src/main/java/bookstore/service/order/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    public List<OrderDto> getAllOrders(String email) {
+        return orderMapper.toDtos(orderRepository.findAllByUserEmail(email));
+    }
+
+    @Transactional
+    @Override
+    public OrderDto setOrderStatus(Long id, String status) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can not find order by id:" + id));
```
Comment (MAMentorFX) [inline src/main/java/bookstore/service/order/impl/OrderServiceImpl.java:43]:
```suggestion
                        "Can not find order by id: " + id));
```

File: src/main/java/bookstore/controller/OrderController.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+@RestController
+@RequestMapping("/orders")
+@RequiredArgsConstructor
+public class OrderController {
+    private OrderService orderService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/{id}")
+    public OrderDto updateStatus(@PathVariable Long id, @RequestBody String status) {
```
Comment (MAMentorFX) [inline src/main/java/bookstore/controller/OrderController.java:27]: Controller may easily deserialize into Enum
```suggestion
    public OrderDto updateStatus(@PathVariable Long id, @RequestBody Status status) {
```

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/bookstore/controller/OrderController.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/orders")
+@RequiredArgsConstructor
+public class OrderController {
+    private OrderService orderService;
+
+    @PreAuthorize("hasRole('ADMIN')")
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:26]: Please add `@PreAuthorize` for other endpoints

File: src/main/java/bookstore/controller/OrderController.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @PutMapping("/{id}")
+    public OrderDto updateStatus(@PathVariable Long id, @RequestBody Order.Status status) {
+        return orderService.setOrderStatus(id, status);
+    }
+
+    @PostMapping
+    public OrderDto placeAnOrder(@RequestBody String shippingAddress,
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:33]: It's better to create CreateOrderRequestDto and validate parameters

File: src/main/java/bookstore/controller/OrderController.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+@RestController
+@RequestMapping("/orders")
+@RequiredArgsConstructor
+public class OrderController {
+    private OrderService orderService;
+
+    @PreAuthorize("hasRole('ADMIN')")
+    @PutMapping("/{id}")
+    public OrderDto updateStatus(@PathVariable Long id, @RequestBody Order.Status status) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:28]: See comment below 
UpdateOrderDto with validation

File: src/main/java/bookstore/controller/OrderController.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    @PostMapping
+    public OrderDto placeAnOrder(@RequestBody String shippingAddress,
+                                 Authentication authentication) {
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        return orderService.saveOrder(userDetails.getUsername(), shippingAddress);
+    }
+
+    @GetMapping
+    public List<OrderDto> getAllOrders(Authentication authentication) {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:40]:
```suggestion
    public List<OrderDto> getAllOrders(Authentication authentication, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/bookstore/controller/OrderController.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        return orderService.getAllOrders(userDetails.getUsername());
+    }
+
+    @GetMapping(value = "/{orderId}/items")
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:45]:
```suggestion
    @GetMapping("/{orderId}/items")
```

File: src/main/java/bookstore/controller/OrderController.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    public List<OrderItemDto> getItemsFromOrder(@PathVariable Long orderId) {
+        return orderService.getItems(orderId);
+    }
+
+    @GetMapping(value = "/{orderId}/items/{itemId}")
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/controller/OrderController.java:50]:
```suggestion
    @GetMapping("/{orderId}/items/{itemId}")
```

File: src/main/java/bookstore/dto/order/OrderItemDto.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.SQLRestriction;
+
+@Entity
+@Table(name = "order_items")
+@SQLDelete(sql = "UPDATE order_items SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "WHERE is_deleted=false")
+@Getter
+@Setter
+public class OrderItemDto {
```
Comment (Elena-Bruyako) [inline src/main/java/bookstore/dto/order/OrderItemDto.java:26]: It looks like Entity, not DTO. Make changes

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's improve your solution!

File: src/main/java/bookstore/model/OrderItem.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    private Book book;
+    @Column(nullable = false)
+    private int quantity;
+    @Column(nullable = false)
+    private BigDecimal price;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    public OrderItem() {
+    }
```
Comment (okuzan) [inline src/main/java/bookstore/model/OrderItem.java:41]: noArgsConstructor

File: src/main/java/bookstore/model/OrderItem.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+    public OrderItem(CartItem cartItem) {
+        this.book = cartItem.getBook();
+        this.price = cartItem.getBook().getPrice();
+        this.quantity = cartItem.getQuantity();
```
Comment (okuzan) [inline src/main/java/bookstore/model/OrderItem.java:46]: `this`s are redundant

File: src/main/java/bookstore/service/order/impl/OrderServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+        return orderItemMapper.toDtos(order.getOrderItems());
+    }
+
+    @Override
+    public OrderItemDto getById(Long orderId, Long itemId) {
+        return orderItemMapper.toDto(orderItemRepository.getById(orderId, itemId).orElseThrow(
+                () -> new EntityNotFoundException(
+                        "Can not find order by id:" + itemId)));
+    }
```
Comment (okuzan) [inline src/main/java/bookstore/service/order/impl/OrderServiceImpl.java:88]: ?
Your entity is uniquely identified by its id field. You don't need to accept orderId and pass it to the repository. Given item.id, you can get its order_id if you want to, it's a foreign key on the entity. Fix this service method and remove repo method

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by sarakhmen — CHANGES_REQUESTED** [review]
Please fix your solution following mentor's advice

File: src/main/java/bookstore/model/OrderItem.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    @ManyToOne
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
+    @Column(nullable = false)
+    private int quantity;
+    @Column(nullable = false)
+    private BigDecimal price;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
```
Comment (kerrrusha) [inline src/main/java/bookstore/model/OrderItem.java:41]:
```suggestion
```

**Review by koval-kn — CHANGES_REQUESTED** [review]

**Review by MAMentorFX — APPROVED** [review]

**Pull request merged**